File: src/utils/Encoding.cpp
#include "Encoding.h"
#include <fstream>

#define WIN32_LEAN_AND_MEAN
#include <windows.h>

namespace Encoding {

std::wstring utf8ToWide(const std::string& str) {
    if (str.empty()) return L"";
    
    int wideLen = MultiByteToWideChar(CP_UTF8, 0, str.c_str(), 
                                       static_cast<int>(str.size()), nullptr, 0);
    if (wideLen == 0) return L"";
    
    std::wstring wide(wideLen, 0);
    MultiByteToWideChar(CP_UTF8, 0, str.c_str(), 
                        static_cast<int>(str.size()), wide.data(), wideLen);
    return wide;
}

std::string wideToUtf8(const std::wstring& wstr) {
    if (wstr.empty()) return "";
    
    int utf8Len = WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), 
                                       static_cast<int>(wstr.size()), 
                                       nullptr, 0, nullptr, nullptr);
    if (utf8Len == 0) return "";
    
    std::string utf8(utf8Len, 0);
    WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), 
                        static_cast<int>(wstr.size()), 
                        utf8.data(), utf8Len, nullptr, nullptr);
    return utf8;
}

std::string shiftJisToUtf8(const std::vector<char>& data) {
    if (data.empty()) return "";
    
    // Convert SHIFT-JIS (codepage 932) to UTF-16
    int wideLen = MultiByteToWideChar(932, 0, data.data(), 
                                       static_cast<int>(data.size()), nullptr, 0);
    if (wideLen == 0) return "";
    
    std::wstring wide(wideLen, 0);
    MultiByteToWideChar(932, 0, data.data(), 
                        static_cast<int>(data.size()), wide.data(), wideLen);
    
    // Convert UTF-16 to UTF-8
    return wideToUtf8(wide);
}

std::string shiftJisToUtf8(const std::string& data) {
    return shiftJisToUtf8(std::vector<char>(data.begin(), data.end()));
}

std::string readFileAsUtf8(const std::string& filePath) {
    std::ifstream file(filePath, std::ios::binary);
    if (!file) return "";
    
    // Read entire file into buffer
    std::vector<char> buffer((std::istreambuf_iterator<char>(file)),
                              std::istreambuf_iterator<char>());
    
    // Convert from SHIFT-JIS to UTF-8
    return shiftJisToUtf8(buffer);
}

} // namespace Encoding

File: rust/src/utils/encoding.rs
//! Encoding utilities for converting SHIFT-JIS to UTF-8.
//!
//! Uses encoding_rs crate instead of Windows APIs for portability.

use encoding_rs::SHIFT_JIS;
use std::io::Read;

/// Convert SHIFT-JIS encoded bytes to UTF-8 string.
pub fn shift_jis_to_utf8(data: &[u8]) -> String {
    let (decoded, _, had_errors) = SHIFT_JIS.decode(data);
    if had_errors {
        // Log warning but continue - some bytes may not be valid SHIFT-JIS
    }
    decoded.into_owned()
}

/// Read a file with SHIFT-JIS encoding and return as UTF-8 string.
pub fn read_file_as_utf8(file_path: &str) -> std::io::Result<String> {
    let mut file = std::fs::File::open(file_path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;
    Ok(shift_jis_to_utf8(&buffer))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_utf8_passthrough() {
        // Pure ASCII/UTF-8 should pass through unchanged
        let data = b"Hello, World!";
        let result = shift_jis_to_utf8(data);
        assert_eq!(result, "Hello, World!");
    }

    #[test]
    fn test_shift_jis_conversion() {
        // SHIFT-JIS encoding of "日本語" (Japanese text)
        let data: &[u8] = &[0x93, 0xFA, 0x96, 0x7B, 0x8C, 0xEA];
        let result = shift_jis_to_utf8(data);
        assert_eq!(result, "日本語");
    }
}
